src/source.cpp

```diff
@@ -430,15 +517,30 @@ MeshSource::MeshSource(pugi::xml_node node)
 
 SourceSite MeshSource::sample(uint64_t* seed) const
 {
-  // sample location and element from mesh
-  auto mesh_location = space_->sample_mesh(seed);
-
-  // Sample source for the chosen element
-  int32_t element = mesh_location.first;
-  SourceSite site = source(element)->sample(seed);
+  // Sample the CDF defined in initialization above
+  int32_t element = space_->sample_element_index(seed);
+
+  // Sample position and apply rejection on spatial domains
+  Position r;
+  while (true) {
+    r = space_->mesh()->sample_element(element, seed);
+    if (this->satisfies_spatial_constraints(r)) {
+      break;
+    }
+  }
 
-  // Replace spatial position with the one already sampled
-  site.r = mesh_location.second;
+  SourceSite site;
+  while (true) {
```

Comment:
Again `do/while` is a natural, beautiful way to handle this type of loop.

Reply:
It is, but when I tried this, clang-format put the closing semicolon on a line by itself, which looks both ugly and confusing to me. Sooo, I'm going to leave this one as is (I will use `do/while` above though)